src/governance-validators.cpp

```diff
@@ -0,0 +1,349 @@
+    : strData(),
+      objJSON(UniValue::VOBJ),
+      fJSONValid(false),
+      strErrorMessages()
+{
+    if(!strDataHexIn.empty()) {
+        SetHexData(strDataHexIn);
+    }
+}
+
+void CProposalValidator::Clear()
+{
+    strData = std::string();
+    objJSON = UniValue(UniValue::VOBJ);
+    fJSONValid = false;
+    strErrorMessages = std::string();
+}
+
+void CProposalValidator::SetHexData(const std::string& strDataHexIn)
+{
+    std::vector<unsigned char> v = ParseHex(strDataHexIn);
+    strData = std::string(v.begin(), v.end());
+    ParseJSONData();
+}
+
+bool CProposalValidator::Validate()
+{
+    if(!ValidateJSON()) {
+        return false;
+    }
+    if(!ValidateName()) {
+        strErrorMessages += "Invalid name;";
+        return false;
+    }
+    if(!ValidateStartEndEpoch()) {
+        strErrorMessages += "Invalid start:end range;";
+        return false;
+    }
+    if(!ValidatePaymentAmount()) {
+        strErrorMessages += "Invalid payment amount;";
+        return false;
+    }
+    if(!ValidatePaymentAddress()) {
+        strErrorMessages += "Invalid payment address;";
+        return false;
+    }
+    if(!ValidateURL()) {
+        strErrorMessages += "Invalid URL;";
+        return false;
+    }
+    return true;
+}
+
+bool CProposalValidator::ValidateJSON()
+{
+    return fJSONValid;
+}
+
+bool CProposalValidator::ValidateName()
+{
+    std::string strName;
+    if(!GetDataValue("name", strName)) {
+        strErrorMessages += "name field not found;";
+        return false;
+    }
+
+    std::string strNameStripped = StripWhitespace(strName);
+
+    if(strNameStripped.empty()) {
+        strErrorMessages += "name is empty;";
+        return false;
+    }
+
+    static const std::string strAllowedChars = "-_abcdefghijklmnopqrstuvwxyz012345789";
+
+    std::transform(strName.begin(), strName.end(), strName.begin(), ::tolower);
+
+    if(strName.find_first_not_of(strAllowedChars) != std::string::npos) {
+        strErrorMessages += "name contains invalid characters;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidateStartEndEpoch()
+{
+    int64_t nStartEpoch = 0;
+    int64_t nEndEpoch = 0;
+
+    if(!GetDataValue("start_epoch", nStartEpoch)) {
+        strErrorMessages += "start_epoch field not found;";
+        return false;
+    }
+
+    if(!GetDataValue("end_epoch", nEndEpoch)) {
+        strErrorMessages += "end_epoch field not found;";
+        return false;
+    }
+
+    if(nEndEpoch <= nStartEpoch) {
+        strErrorMessages += "end_epoch <= start_epoch field not found;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidatePaymentAmount()
+{
+    double dValue = 0.0;
+
+    if(!GetDataValue("payment_amount", dValue)) {
+        strErrorMessages += "payment_amount field not found;";
+        return false;
+    }
+
+    if(dValue <= 0.0) {
+        strErrorMessages += "payment_amount invalid;";
```

Comment:
>So I'm not sure how to implement this.
1. Estimate blocks height at `end_epoch` using `consensus.nPowTargetSpacing`, current timestamp and current block height;
2. Find next sb for that height, same way as in `getgovernanceinfo` - probably can move calculation to a standalone function e.g.
```
void CSuperblock::GetNearestSuperblocksHeights(int nBlockHeight, int &nLastSuperblock, int &nNextSuperblock)
{
    const Consensus::Params& consensusParams = Params().GetConsensus();
    int nSuperblockStartBlock = consensusParams.nSuperblockStartBlock;
    int nSuperblockCycle = consensusParams.nSuperblockCycle;

    // Get first superblock
    int nFirstSuperblockOffset = (nSuperblockCycle - nSuperblockStartBlock % nSuperblockCycle) % nSuperblockCycle;
    int nFirstSuperblock = nSuperblockStartBlock + nFirstSuperblockOffset;

    if(nBlockHeight < nFirstSuperblock) {
        nLastSuperblock = 0;
        nNextSuperblock = nFirstSuperblock;
    } else {
        nLastSuperblock = nBlockHeight - nBlockHeight % nSuperblockCycle;
        nNextSuperblock = nLastSuperblock + nSuperblockCycle;
    }
}
```
and then just reuse it in both cases;
3. Use `CAmount CSuperblock::GetPaymentsLimit(int nBlockHeight)` to get max total budget at `nNextSuperblock`.